misc/architecture-model-generator/src/main/java/io/ballerina/architecturemodelgenerator/generators/entity/EntityModelGenerator.java

```diff
@@ -222,15 +259,54 @@ private String getSelfCardinality(TypeSymbol typeSymbol, String entityName) {
 
     private String getAssociateCardinality(boolean isArray, boolean isOptional, boolean isNillable) {
         // todo: double check
-        if (isArray && !isOptional && !isNillable) {
-            return CardinalityValue.ONE_OR_MANY.getValue();
-        } else if (isArray) {
+        if (isArray) {
             return CardinalityValue.ZERO_OR_MANY.getValue();
-        } else if (isOptional || isNillable) {
-            return CardinalityValue.ZERO_OR_ONE.getValue();
         } else {
-            return CardinalityValue.ONE_AND_ONLY_ONE.getValue();
+            if (isOptional || isNillable) {
+                return CardinalityValue.ZERO_OR_ONE.getValue();
+            } else {
+                return CardinalityValue.ONE_AND_ONLY_ONE.getValue();
+            }
+        }
```

Comment:
IMO better to improve the comments/javadocs in here, as we making opinionated decisions in some cases